service/history/historyEngine.go

```diff
@@ -151,6 +151,8 @@ var (
 	ErrConsistentQueryNotEnabled = &workflow.BadRequestError{Message: "cluster or domain does not enable strongly consistent query but strongly consistent query was requested"}
 	// ErrConsistentQueryBufferExceeded is error indicating that too many consistent queries have been buffered and until buffered queries are finished new consistent queries cannot be buffered
 	ErrConsistentQueryBufferExceeded = &workflow.InternalServiceError{Message: "consistent query buffer is full, cannot accept new consistent queries"}
+	// ErrConcurrentStartRequest is error indicating there is an outstanding start workflow request. The incoming request fails to acquires the lock before the outstanding request finishes.
+	ErrConcurrentStartRequest = &workflow.ServiceBusyError{Message: "an outstanding start workflow request is in-progress. Failed to acquire the resource."}
```

Comment:
The error message is not very accurate I think. It's not necessarily that there's another outstanding start workflow request when the timeout happens. The timeout can also happen if customer starts a workflow and schedules a large number of activities, then the lock will be held by all the RecordActivityStarted API calls, results in timeout for the StartWorkflowAPI call.